simpeg/directives/_regularization.py

```diff
@@ -0,0 +1,482 @@
+from __future__ import annotations
+
+import warnings
+
+import numpy as np
+from dataclasses import dataclass
+from typing import Literal
+from .directives import InversionDirective, UpdatePreconditioner, BetaSchedule
+from simpeg.regularization import Sparse, BaseSparse, SmoothnessFirstOrder
+from simpeg.utils import validate_integer, validate_float
+from simpeg.utils.code_utils import deprecate_property
+
+
+@dataclass
+class IRLSMetrics:
+    """
+    Data class to store metrics used by the IRLS algorithm.
+
+    Parameters
+    ----------
+
+    input_norms: List of norms temporarily stored during the initialization.
+    irls_iteration_count: Number of IRLS iterations.
+    start_irls_iter: Iteration number when the IRLS process started.
+    f_old: Previous value of the regularization function.
+    mode: Mode of the IRLS algorithm, 1 for l2-norm and 2 for lp-norm iterations.
+    """
+
+    input_norms: list[float]
+    irls_iteration_count = 0
+    start_irls_iter = 0
+    f_old: float = 0.0
+    mode: Literal[1, 2] = 1
+
+
+class UpdateIRLS(InversionDirective):
+    """
+    Directive to control the IRLS iterations for regularization.Sparse.
+
+    Parameters
+    ----------
+
+    beta_search: bool
+        Proceed with a beta search step if the target misfit is outside the tolerance.
```

Comment:
`beta_search` is not an argument of the constructor. Should docstrings be updated or the constructor itself?